rp2040-hal/src/xosc.rs

```diff
@@ -0,0 +1,192 @@
+//! Crystal Oscillator (XOSC)
+// See [Chapter 2 Section 16](https://datasheets.raspberrypi.org/rp2040/rp2040_datasheet.pdf) for more details
+
+use core::{
+    convert::Infallible,
+    ops::RangeInclusive
+};
+
+use embedded_time::{
+    fraction::Fraction,
+    fixed_point::FixedPoint,
+    rate::{
+        Hertz,
+        Megahertz,
+    },
+    duration::{
+        Seconds,
+        Milliseconds,
+        Duration
+    }
+};
+
+use nb::Error::WouldBlock;
+
+/// State of the Crystal Oscillator (typestate trait)
+pub trait State {}
+
+/// XOSC is disabled (typestate)
+pub struct Disabled;
+
+/// XOSC is initialized, ie we've given parameters (typestate)
+pub struct Initialized {
+    freq_hz: Hertz<u64>
+}
+
+/// Stable state (typestate)
+pub struct Stable{
+    freq_hz: Hertz<u64>
+}
+
+/// XOSC is disabling (typestate)
+pub struct Disabling;
+
+/// XOSC is in dormant mode (see Chapter 2, Section 16, §5)
+pub struct Dormant;
+
+impl State for Disabled {}
+impl State for Initialized {}
+impl State for Stable {}
+impl State for Disabling {}
+impl State for Dormant {}
+
+/// Possible errors when initializing the CrystalOscillator
+pub enum Error {
+    /// Frequency is out of the 1-15MHz range (see datasheet)
+    FrequencyOutOfRange,
+
+    /// Argument is bad : overflows, ...
+    BadArgument
+}
+
+/// A Crystal Oscillator.
+pub struct CrystalOscillator<S: State> {
+    device: rp2040_pac::XOSC,
+    state: S
+}
+
+impl<S: State> CrystalOscillator<S> {
+    /// Transitions the oscillator to another state.
+    fn transition<To: State>(self, state: To) -> CrystalOscillator<To> {
+        CrystalOscillator {
+            device: self.device,
+            state: state
+        }
+    }
+
+    /// Releases the underlying device.
+    pub fn free(self) -> rp2040_pac::XOSC {
+        self.device
+    }
+}
+
+impl CrystalOscillator<Disabled> {
+
+    /// Creates a new CrystalOscillator from the underlying device.
+    pub fn new(dev: rp2040_pac::XOSC) -> Self {
+        CrystalOscillator {
+            device: dev,
+            state: Disabled
+        }
+    }
+
+    /// Initializes the XOSC : frequency range is set, startup delay is calculated and set.
+    pub fn initialize(self, frequency: Megahertz<u32>) -> Result<CrystalOscillator<Initialized>, Error> {
+
+        const ALLOWED_FREQUENCY_RANGE: RangeInclusive<Megahertz<u32>> = Megahertz(1)..=Megahertz(15);
+        const STABLE_DELAY: Milliseconds<u64> = Milliseconds(1_u64);
+        const DIVIDER: Fraction = Fraction::new(1, 256);
+
+        if !ALLOWED_FREQUENCY_RANGE.contains(&frequency) {
+            return Err(Error::FrequencyOutOfRange)
+        }
+
+        self.device.ctrl.write(|w| {
+            w.freq_range()._1_15mhz();
+            w
+        });
+
+        let freq_hz: Hertz<u64> = frequency.into();
+        let delay_sec: Seconds<u64> = STABLE_DELAY.into();
```

Comment:
Looking again, this might truncate the sub-seconds so this could come out to being 0s.

Reply:
You were right ! I struggled for an hour to try and write a unit test within the crate but apparently it's not possible when the crate is no_std. None of the tricks I found online worked so here is the code of the test, that I ran on a standalone project, if you happen to know how to run it from within this crate : 

```rust
#[cfg(test)]
mod tests {
use std::convert::TryInto;
use embedded_time::{
        fraction::Fraction,
        fixed_point::FixedPoint,
        rate::{
            Hertz,
            Megahertz,
            Rate
        },
        duration::{
            Seconds,
            Milliseconds,
            Duration,
            Generic,
        }
    };
    #[test]
    fn calculate_xosc_delay() {
        let frac: Fraction = Fraction::new(256, 1);
        let milliseconds = Milliseconds(1_u32);
        let seconds: Seconds = milliseconds.into();
        let frequency: Hertz = Hertz(12_000_000_u32);
        
        let delay_to_hz: Hertz = milliseconds.to_rate().unwrap();
        assert_eq!(*delay_to_hz.integer(), 1000);

        let oper = frequency.checked_div(delay_to_hz.integer()).and_then(|r| {
            r.to_generic::<u32>(frac).ok()
        });
        assert!(oper.is_some());

        let res = oper.unwrap();

        let to_u16: Result<u16, _> = (*res.integer()).try_into();
        assert!(to_u16.is_ok());

        //Value from datasheet (Chapter 2, Section 16, §3)
        assert_eq!(to_u16.unwrap()+1, 47);
    }
}
```

Reply:
Oh yes I remember I had the same problem. The tests need to be compiled for the host instead of for the pico. But I don't remember how I did this last time.

I would try to move the calculation to its own function and then add this unit test to the HAL.